core/trino-main/src/main/java/io/trino/operator/aggregation/DecimalAverageAggregation.java

```diff
@@ -114,119 +119,131 @@ public AggregationMetadata specialize(BoundSignature boundSignature)
                 ImmutableList.of(new AccumulatorStateDescriptor<>(
                         stateInterface,
                         stateSerializer,
-                        new LongDecimalWithOverflowAndLongStateFactory())));
+                        new Int128StateFactory()),
+                        new AccumulatorStateDescriptor<>(
+                                LongState.class,
+                                StateCompiler.generateStateSerializer(LongState.class),
+                                StateCompiler.generateStateFactory(LongState.class)),
+                        new AccumulatorStateDescriptor<>(
+                                NullableLongState.class,
+                                StateCompiler.generateStateSerializer(NullableLongState.class),
+                                StateCompiler.generateStateFactory(NullableLongState.class))));
     }
 
-    public static void inputShortDecimal(LongDecimalWithOverflowAndLongState state, Block block, int position)
+    public static void inputShortDecimal(Int128State decimalState, LongState counterState, NullableLongState overflowState, Block block, int position)
     {
-        state.addLong(1); // row counter
+        long[] decimal = decimalState.getArray();
+        int decimalOffset = decimalState.getArrayOffset();
 
-        state.setNotNull();
-
-        long[] decimal = state.getDecimalArray();
-        int offset = state.getDecimalArrayOffset();
+        decimalState.setNotNull(true);
+        counterState.setValue(counterState.getValue() + 1);
 
         long rightLow = block.getLong(position, 0);
         long rightHigh = rightLow >> 63;
 
         long overflow = addWithOverflow(
-                decimal[offset],
-                decimal[offset + 1],
+                decimal[decimalOffset],
+                decimal[decimalOffset + 1],
                 rightHigh,
                 rightLow,
                 decimal,
-                offset);
+                decimalOffset);
 
-        state.addOverflow(overflow);
+        overflow += overflowState.getValue();
+        overflowState.setNull(overflow == 0 & overflowState.isNull());
```

Comment:
why not just `overflowState.setNull(overflow == 0)`?

Reply:
It is not necessary. It was done to do not change the previouse behaviour. Generally, it does not matter but you right.